File: python/cache/transport.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Dict, Optional


@dataclass
class ReplicationMessage:
    op: str
    key: Optional[str] = None
    value: object = None
    ttl_ms: Optional[int] = None
    expires_at: Optional[int] = None
    version: Optional[int] = None
    channel: Optional[str] = None
    payload: object = None


class Transport:
    def send(self, peer_id: str, message: ReplicationMessage) -> None:
        raise NotImplementedError


class InMemoryTransport(Transport):
    def __init__(self) -> None:
        self._handlers: Dict[str, Callable[[ReplicationMessage], None]] = {}

    def register(self, peer_id: str, handler: Callable[[ReplicationMessage], None]) -> None:
        self._handlers[peer_id] = handler

    def send(self, peer_id: str, message: ReplicationMessage) -> None:
        handler = self._handlers.get(peer_id)
        if handler:
            handler(message)

File: go/cache/transport.go
package cache

import "sync"

type ReplicationMessage struct {
	Op        string
	Key       string
	Value     any
	ExpiresAt *int64
	Version   uint64
	Channel   string
	Payload   any
}

type ReplicationHandler func(message ReplicationMessage)

type Transport interface {
	Send(peerID string, message ReplicationMessage) error
}

type InMemoryTransport struct {
	mu       sync.RWMutex
	handlers map[string]ReplicationHandler
}

func NewInMemoryTransport() *InMemoryTransport {
	return &InMemoryTransport{
		handlers: map[string]ReplicationHandler{},
	}
}

func (t *InMemoryTransport) Register(peerID string, handler ReplicationHandler) {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.handlers[peerID] = handler
}

func (t *InMemoryTransport) Send(peerID string, message ReplicationMessage) error {
	t.mu.RLock()
	handler := t.handlers[peerID]
	t.mu.RUnlock()
	if handler != nil {
		handler(message)
	}
	return nil
}
